### PR TITLE
 feat: Support structured JSON logging for controller, executor, and server

### PR DESCRIPTION
**Pull request by terrytangyuan**
This option makes the log easier to parse and aggregate by services like Stackdriver, Splunk, and Logstash. 

Signed-off-by: terrytangyuan <terrytangyuan@gmail.com>

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

### CODE REVIEW

**Review by simster7 — CHANGES_REQUESTED** [review]
Can this be added to the server as well? Maybe even the CLI too?

**Comment by terrytangyuan** [comment]
> Can this be added to the server as well? Maybe even the CLI too?

I can see this can be useful for the server (just added to the server as well) but probably not for the CLI since there doesn't seem to be any use cases to parse logs from the CLI.

**Comment by simster7** [comment]
> I can see this will be useful for the server (just added to the server as well) but probably not for the CLI since there doesn't seem to be any use case to parse logs from the CLI.

The CLI can be used in an automated/scripted pipeline such as Jenkins. Getting logs from those pipelines is certainly a useful use case.

File: docs/cli/argo_server.md [hunk 1]
```diff
@@ -29,6 +29,7 @@ See https://argoproj.github.io/argo-workflows/argo-server.md
       --event-worker-count int               how many event workers to run (default 4)
   -h, --help                                 help for server
       --hsts                                 Whether or not we should add a HTTP Secure Transport Security header. This only has effect if secure is enabled. (default true)
+      --json-log-format                      Whether to use JSON formatter for logs. Default to use plain text formatter
```
Comment (alexec) [inline docs/cli/argo_server.md:32]: I'd prefer `--log-format=json`

Comment (alexec) [reply docs/cli/argo_server.md:32]: does Argo CD do something similar at all?

Comment (terrytangyuan) [reply docs/cli/argo_server.md:32]: Did a quick search - looks like it does https://github.com/argoproj/argo-cd/blob/a96b476f167b08ee900887b7f9e026a06c45e899/util/cli/cli.go#L145-L156

It uses `--logformat` [here](https://github.com/argoproj/argo-cd/blob/7af5837ffabd5420b52b7980baeab19f046f9294/cmd/argocd-server/commands/argocd_server.go#L144). I think we should do `--log-format` with `text` being the default here.

**Comment by terrytangyuan** [comment]
> > I can see this will be useful for the server (just added to the server as well) but probably not for the CLI since there doesn't seem to be any use case to parse logs from the CLI.
> 
> The CLI can be used in an automated/scripted pipeline such as Jenkins. Getting logs from those pipelines is certainly a useful use case.

Sounds good. I'll refactor and add this to other commands as well.

**Comment by terrytangyuan** [comment]
Comments addressed. Please take another look when you get a chance.

**Comment by terrytangyuan** [comment]
@alexec @simster7 Would you like to take another look at this when you get a chance? It would be great to get this in before next rc release.

**Comment by alexec** [comment]
I’m not sure about this.  Most CLI commands do not do any logging, so I’m not clear of what problem we’re trying to solve?

**Comment by terrytangyuan** [comment]
> I’m not sure about this. Most CLI commands do not do any logging, so I’m not clear of what problem we’re trying to solve?

@simster7 What do you think? Shall we just add this flag for controller & executor & server? 

@alexec Are you okay with adding this for these three commands though?

**Comment by alexec** [comment]
I think this would be useful for the three commands you mention.

**Comment by terrytangyuan** [comment]
Thanks for confirming! I just narrowed down the scope of this PR to those three commands. Let me know if I missed anything else.

**Review by alexec — APPROVED** [review]

**Comment by terrytangyuan** [comment]
The failing test seems to be related to master branch and unrelated to this PR.

**Pull request merged**